controllers/dscinitialization/utils.go

```diff
@@ -18,132 +20,180 @@ import (
-			if err != nil {
-				return err
-			}
-		}
+	// Patch monitoring namespace
+	err := r.patchMonitoringNS(ctx, dscInit)
+	if err != nil {
+		log.Error(err, "error patch monitoring namespace for managed cluster")
+		return err
 	}
 
 	// Create default NetworkPolicy for the namespace
-	err = r.reconcileDefaultNetworkPolicy(ctx, name, dscInit, platform)
+	err = r.reconcileDefaultNetworkPolicy(ctx, dscInit, platform)
 	if err != nil {
-		log.Error(err, "error reconciling network policy ", "name", name)
-		return err
+		log.Error(err, "error reconciling network policy ", "name", dscInit.Spec.ApplicationsNamespace)
+		return fmt.Errorf("error: %w", err)
 	}
 
 	// Create odh-common-config Configmap for the Namespace
-	err = r.createOdhCommonConfigMap(ctx, name, dscInit)
+	err = r.createOdhCommonConfigMap(ctx, dscInit)
 	if err != nil {
 		log.Error(err, "error creating configmap", "name", "odh-common-config")
 		return err
 	}
 
 	// Create default Rolebinding for the namespace
-	err = r.createDefaultRoleBinding(ctx, name, dscInit)
+	err = r.createDefaultRoleBinding(ctx, dscInit)
 	if err != nil {
-		log.Error(err, "error creating rolebinding", "name", name)
+		log.Error(err, "error creating rolebinding", "name", dscInit.Spec.ApplicationsNamespace)
 		return err
 	}
 	return nil
 }
 
-func (r *DSCInitializationReconciler) createDefaultRoleBinding(ctx context.Context, name string, dscInit *dsciv1.DSCInitialization) error {
+func (r *DSCInitializationReconciler) appNamespaceHandler(ctx context.Context, dscInit *dsciv1.DSCInitialization, platform cluster.Platform) error {
 	log := logf.FromContext(ctx)
+
+	nsList := &corev1.NamespaceList{}
+	ns := &corev1.Namespace{}
+	dsciNsName := dscInit.Spec.ApplicationsNamespace
+
+	if err := r.List(ctx, nsList, client.MatchingLabels{
+		labels.CustomizedAppNamespace: labels.True,
+	}); err != nil {
+		return err
+	}
+
+	switch len(nsList.Items) {
+	case 0:
+		// create namespace if not exist
+		desiredAppNS := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: dsciNsName,
+			},
+		}
+		if err := r.Get(ctx, client.ObjectKeyFromObject(desiredAppNS), ns); err != nil {
+			if !k8serr.IsNotFound(err) {
+				return err
+			}
+		}
+		log.Info("Application namespace set in DSCI not found, creating it with labels", "name", dsciNsName)
+		// // ensure generatedd-namespace:true and security label always on it
+		return r.createAppNamespace(ctx, dsciNsName, platform, map[string]string{labels.ODH.OwnedNamespace: labels.True}) // this indicate when uninstall, namespace will be deleted
+	case 1:
+		if nsList.Items[0].Name != dsciNsName {
+			return errors.New("DSCI must used the same namespace which has opendatahub.io/application-namespace=true label")
+		}
+		// ensure appliation-namespace:true and security label always on it
+		return r.createAppNamespace(ctx, dsciNsName, platform, map[string]string{labels.CustomizedAppNamespace: labels.True})
```

Comment:
nit: is this additional map needed ? my understating is that in this case, the namespace carries this label already.

Reply:
yes, not necessary, will remove it.